e2e/testsuite/tx.go

```diff
@@ -33,7 +33,48 @@ import (
 
 // BroadcastMessages broadcasts the provided messages to the given chain and signs them on behalf of the provided user.
 // Once the broadcast response is returned, we wait for a few blocks to be created on both chain A and chain B.
-func (s *E2ETestSuite) BroadcastMessages(ctx context.Context, chain ibc.Chain, user ibc.Wallet, msgs ...sdk.Msg) sdk.TxResponse {
+func (s *E2ETestSuite) BroadcastMessages(ctx context.Context, chainA ibc.Chain, user ibc.Wallet, chainB ibc.Chain, msgs ...sdk.Msg) sdk.TxResponse {
+	cosmosChain, ok := chainA.(*cosmos.CosmosChain)
+	if !ok {
+		panic("BroadcastMessages expects a cosmos.CosmosChain")
+	}
+
+	broadcaster := cosmos.NewBroadcaster(s.T(), cosmosChain)
+
+	// strip out any fields that may not be supported for the given chain version.
+	msgs = sanitize.Messages(cosmosChain.Nodes()[0].Image.Version, msgs...)
+
+	broadcaster.ConfigureClientContextOptions(func(clientContext client.Context) client.Context {
+		// use a codec with all the types our tests care about registered.
+		// BroadcastTx will deserialize the response and will not be able to otherwise.
+		cdc := Codec()
+		return clientContext.WithCodec(cdc).WithTxConfig(authtx.NewTxConfig(cdc, []signingtypes.SignMode{signingtypes.SignMode_SIGN_MODE_DIRECT}))
+	})
+
+	broadcaster.ConfigureFactoryOptions(func(factory tx.Factory) tx.Factory {
+		return factory.WithGas(DefaultGasValue)
+	})
+
+	// Retry the operation a few times if the user signing the transaction is a relayer. (See issue #3264)
+	var resp sdk.TxResponse
+	var err error
+	broadcastFunc := func() (sdk.TxResponse, error) {
+		return cosmos.BroadcastTx(ctx, broadcaster, user, msgs...)
+	}
+	if s.relayers.ContainsRelayer(s.T().Name(), user) {
+		// Retry five times, the value of 5 chosen is arbitrary.
+		resp, err = s.retryNtimes(broadcastFunc, 5)
+	} else {
+		resp, err = broadcastFunc()
+	}
+	s.Require().NoError(err)
+
+	s.Require().NoError(test.WaitForBlocks(ctx, 2, chainA, chainB))
+
+	return resp
+}
+
+func (s *E2ETestSuite) BroadcastMessagesWithoutChainAB(ctx context.Context, chain ibc.Chain, user ibc.Wallet, msgs ...sdk.Msg) sdk.TxResponse {
```

Comment:
It means that the function does not need to pass both chainA and chainB parameters. Initially, my idea was to keep the existing functions because they were being used for UpgradeTest and GrandpaTest. I will consider bringing it together into a single function in the next work.